### PR TITLE
JSON encoding / decoding customization for JSON based cass

### PR DESCRIPTION
**Pull request by taylorotwell**
This PR allows the customization of `json_encode` and `json_decode` for the `array` and `collection` casts, as well as `AsArrayObject`, `AsCollection`, `AsEnumArrayObject`, etc.

```php
use Illuminate\Database\Eloquent\Casts\Json;

Json::encodeUsing(fn ($value) => json_encode($value));

Json::decodeUsing(fn ($value, $associative) => json_decode($value, $associative));
```

This customization would typically be done in a service provider.

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Casts/Json.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+     * @var callable|null
+     */
+    protected $encoder;
+
+    /**
+     * The custom JSON decode.
+     *
+     * @var callable|null
+     */
+    protected $decoder;
```
Comment (driesvints) [inline src/Illuminate/Database/Eloquent/Casts/Json.php:19]: @taylorotwell I guess these need to become static?

File: src/Illuminate/Database/Eloquent/Casts/Json.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Illuminate\Database\Eloquent\Casts;
+
+class Json
```
Comment (gocanto) [inline src/Illuminate/Database/Eloquent/Casts/Json.php:5]: May we be able to have this in the Support package instead?

Comment (taylorotwell) [reply src/Illuminate/Database/Eloquent/Casts/Json.php:5]: No. That would imply it affects JSON serialization **everywhere**. Having it here demonstrates it is for **casting only**.

**Comment by morloderex** [comment]
What if 2 columns within the same project needs different json encoding / decoding? This customization object seems to be global? Would that require a custom cast?

**Comment by taylorotwell** [comment]
@morloderex I guess that require a custom cast. My thinking was that if you're using something like `JSON_UNESCAPED_UNICODE` you probably want that on all your JSON casts throughout your application.

**Pull request merged**